ZeroMQ->NNG change for thread-safe sockets

eigenraven opened the pull request:
We discussed this a long time ago: ZeroMQ sockets are not thread-safe, so a lot of threads end up creating their own client objects that create connections to other nodes if communication is required. This causes a lot of unnecessary TCP connections to get created, leading to running over default socket limits, it can lead to worse slow start behaviour on those sockets and other similar scaling issues. NNG is a library made by the original authors of 0mq after learning from its mistakes and nanomsg's mistakes.

Crucially, nng sockets are thread-safe and can be used from multiple threads. Even patterns such as req-rep can be done on the same socket from multiple threads by using `nng_context` objects (instead of the req-router-dealer-rep 0mq pattern).

One big difference between nng and 0mq sockets is that send operations will block until a receiving endpoint is connected, rather than going blindly into a queue that might never connect to an endpoint. This broke a few tests which were connecting to non-existent endpoints, so they needed fixing. Some tests also created the sending endpoints and sent messages before creating the corresponding receivers which led to deadlocks. Passing the `NNG_FLAG_NONBLOCK` flag wouldn't solve this issue as in that case the send function returns an error EAGAIN, so some parts of the code (mostly tests) need to be changed to make sure receivers become available at some point during send calls.

To-do:
 - [x] Replace 0MQ with NNG in CMake
 - [x] Switch the message/endpoint classes to wrap NNG sockets&contexts
 - [x] Find instances of thread-local sockets and replace them with global sockets
 - [x] Add a thread-safe concurrent map wrapper to replace manual handling of mutexes+maps, because quite a few of the thread-local maps and also some global maps use the same patterns over and over making it prone to mistakes with forgotten locks.
 - [x] Replace map+mutex usage with the concurrent map type where appropriate

Closes #266

csegarragonz reviewed (approved):
Overall, this looks great! Thanks for taking the time to do this refactor. The thread-unsafety issue in ZeroMQ had caused many headaches in the past, so hopefully this fixes them.

I only have one minor comment: are all methods in the new (and very cool) `ConcurrentMap` class used? I feel some of them are not, but I can not be entirely sure. I'm reluctant to add methods that we don't use.

Other than that, just left some minor comments around.

csegarragonz commented:
Also, @eigenraven does this close #202?

eigenraven commented on include/faabric/transport/Message.h:
```diff
@@ -26,42 +44,89 @@ enum MessageResponseCode
+    std::span<uint8_t> udata();
+    std::span<const uint8_t> udata() const;
+
+    std::vector<uint8_t> dataCopy() const;
```
It specifically returns a copy rather than a reference to the data, so it must use `std::vector` to retain ownership of the allocated bytes.

eigenraven commented on include/faabric/transport/Message.h:
```diff
@@ -26,42 +44,89 @@ enum MessageResponseCode
 
   private:
-    std::vector<uint8_t> buffer;
+    nng_msg* nngMsg = nullptr;
```
The Message class here is the smart pointer implementation for nngMsg

eigenraven commented on src/transport/Message.cpp:
```diff
@@ -1,33 +1,65 @@
+Message::Message(size_t bufferSize)
+{
+    if (int ec = nng_msg_alloc(&nngMsg, bufferSize); ec < 0) {
+        SPDLOG_CRITICAL("Error allocating a message of size {}: {}",
```
`bad_alloc` and allocation failure is generally unrecoverable without carefully coding the entire codebase to guard every allocation and fail gracefully, so I thought critical was appropriate as this is not an error after which execution of the program could continue.

eigenraven commented on src/transport/MessageEndpoint.cpp:
```diff
@@ -191,171 +245,163 @@ zmq::socket_t socketFactory(zmq::socket_type socketType,
-              throw std::runtime_error("Error sending message");
+void MessageEndpoint::close()
+{
+    std::visit(
```
Added, this also just follows the example std::visit usage from <https://en.cppreference.com/w/cpp/utility/variant/visit>.

eigenraven commented on src/transport/MessageEndpoint.cpp:
```diff
@@ -191,171 +245,163 @@ zmq::socket_t socketFactory(zmq::socket_type socketType,
+
+    int ec = nng_close(socket);
+    if (ec != 0) {
+        SPDLOG_WARN("Error when closing socket: {}", nng_strerror(ec));
```
The only error that nng_close currently returns is ECLOSED which means the socket was already closed, so while probably a programming bug somewhere it is harmless.

eigenraven commented:
> I only have one minor comment: are all methods in the new (and very cool) ConcurrentMap class used? I feel some of them are not, but I can not be entirely sure. I'm reluctant to add methods that we don't use.

I added test methods to cover everything else, they're not all used in faabric right now but I added them to mirror the `std::unordered_map`/`absl::flat_hash_map` interface where it was possible to build a thread-safe wrapper (so e.g. no `operator[]`). There still are other places in faabric&faasm that use maps with mutexes that use the various methods, so this should make porting those over easier.

> does this close https://github.com/faasm/faabric/issues/202?

It makes good progress on it, but I've definitely not covered every place where spans could be used.

csegarragonz commented:
Great, thanks for your answers. LGTM, let's wait for Simon to take a look too.

Shillaker reviewed (approved):
Hello! I've had a quick look and it all looks sensible to me! I haven't had time to look at everything in detail, but I trust you two, so LGTM!

csegarragonz commented:
@eigenraven I have a deadline in six hours, will merge everything in tomorrow.
